server/commands/list_objects.go

```diff
@@ -0,0 +1,370 @@
+	tracer                trace.Tracer
+	meter                 metric.Meter
+	ListObjectsDeadline   time.Duration
+	ListObjectsMaxResults uint32
+	ResolveNodeLimit      uint32
+}
+
+// NewListObjectsQuery creates a ListObjectsQuery
+func NewListObjectsQuery(datastore storage.OpenFGADatastore, tracer trace.Tracer, logger logger.Logger, meter metric.Meter, listObjectsDeadline time.Duration, listObjectsMaxResults uint32, ResolveNodeLimit uint32) *ListObjectsQuery {
+	return &ListObjectsQuery{
+		datastore:             datastore,
+		logger:                logger,
+		tracer:                tracer,
+		meter:                 meter,
+		ListObjectsDeadline:   listObjectsDeadline,
+		ListObjectsMaxResults: listObjectsMaxResults,
+		ResolveNodeLimit:      ResolveNodeLimit,
+	}
+}
+
+// Execute the ListObjectsQuery, returning a list of object IDs
+func (q *ListObjectsQuery) Execute(ctx context.Context, req *openfgapb.ListObjectsRequest) (*openfgapb.ListObjectsResponse, error) {
+	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return nil, err
+	}
+	iter, err := q.getTupleIterator(ctx, req.StoreId, req.Type)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Stop()
+
+	g := new(errgroup.Group)
+	g.SetLimit(MaximumConcurrentChecks)
+
+	uniqueObjectIdsSet := utils.NewConcurrentSet()
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	var cancel context.CancelFunc
+	if q.ListObjectsDeadline != 0 {
+		timeoutCtx, cancel = context.WithTimeout(ctx, q.ListObjectsDeadline)
+		defer cancel()
+	}
+
+	objectIDs := make([]string, 0)
+
+	go func() {
+		var objectsFound uint32
+
+		// iterate over all object IDs in the store and check if the user has relation with each
+		for {
+			if atomic.LoadUint32(&objectsFound) >= q.ListObjectsMaxResults {
+				break
+			}
+
+			t, err := iter.Next()
+			if err != nil {
+				if err == storage.TupleIteratorDone {
+					break
+				} else {
+					errChan <- err
+				}
+			}
+			checkFunction := func() error {
+				_, objectID := tuple.SplitObject(t.Key.Object)
+				if uniqueObjectIdsSet.Has(objectID) {
+					return nil
+				}
+				query := NewCheckQuery(q.datastore, q.tracer, q.meter, q.logger, q.ResolveNodeLimit)
+
+				resp, err := query.Execute(timeoutCtx, &openfgapb.CheckRequest{
+					StoreId:              req.StoreId,
+					AuthorizationModelId: req.AuthorizationModelId,
+					TupleKey: &openfgapb.TupleKey{
+						Object:   t.Key.Object,
+						Relation: req.GetRelation(),
+						User:     req.GetUser(),
+					},
+				})
+				if err != nil {
+					return err
+				}
+
+				if resp.Allowed && !uniqueObjectIdsSet.Has(objectID) && atomic.LoadUint32(&objectsFound) < q.ListObjectsMaxResults {
+					uniqueObjectIdsSet.Add(objectID)
+					atomic.AddUint32(&objectsFound, 1)
+				}
+
+				return nil
+			}
+
+			g.Go(checkFunction)
+		}
+
+		// iterate over all tuples in the contextual tuples input
+		if req.ContextualTuples != nil {
```

Comment:
I'd do these first because it's a pre-defined quantity of routines which (if I remember correctly) is less than the total number of concurrency, so these could be resolved without having to wait on possibly 100 tuples returned on the iterator.

Reply:
This has been refactored